lib/Sema/TypeCheckCompilerEvaluable.cpp

```diff
@@ -26,26 +26,205 @@ using namespace swift;
+  bool CompilerEvaluable = true;
+
+public:
+  CheckCompilerEvaluableBody(TypeChecker &TC,
+                             const AbstractFunctionDecl *CheckingFunc)
+      : TC(TC), CheckingFunc(CheckingFunc) {}
+
+  std::pair<bool, Expr *> walkToExprPre(Expr *E) override {
+    // If this is the ignored part of a DotSyntaxBaseIgnored, then we can accept
+    // it without walking it.
+    if (auto *parentDotSyntaxBaseIgnored =
+            dyn_cast_or_null<DotSyntaxBaseIgnoredExpr>(Parent.getAsExpr()))
+      if (parentDotSyntaxBaseIgnored->getLHS() == E)
+        return {false, E};
+
+    if (!checkCompilerRepresentable(E->getType())) {
+      TC.diagnose(E->getLoc(), diag::compiler_evaluable_forbidden_type,
+                  E->getType())
+          .highlight(E->getSourceRange());
+      CompilerEvaluable = false;
+      return {false, E};
+    }
+
+    switch (E->getKind()) {
+    #define ALWAYS_ALLOWED(ID)                                                 \
+      case ExprKind::ID:                                                       \
+        return {true, E};
+    #define SOMETIMES_ALLOWED(ID)                                              \
+      case ExprKind::ID:                                                       \
+         return checkExpr##ID(cast<ID##Expr>(E));
+
+    ALWAYS_ALLOWED(NilLiteral)
+    ALWAYS_ALLOWED(IntegerLiteral)
+    ALWAYS_ALLOWED(BooleanLiteral)
+    ALWAYS_ALLOWED(MagicIdentifierLiteral)
+    ALWAYS_ALLOWED(DiscardAssignment)
+    SOMETIMES_ALLOWED(DeclRef)
+    ALWAYS_ALLOWED(Type)
+    SOMETIMES_ALLOWED(OtherConstructorDeclRef)
+    ALWAYS_ALLOWED(DotSyntaxBaseIgnored)
+    ALWAYS_ALLOWED(MemberRef)
+    ALWAYS_ALLOWED(Paren)
+    ALWAYS_ALLOWED(DotSelf)
+    ALWAYS_ALLOWED(Try)
+    ALWAYS_ALLOWED(ForceTry)
+    ALWAYS_ALLOWED(OptionalTry)
+    ALWAYS_ALLOWED(Tuple)
+    ALWAYS_ALLOWED(Subscript)
+    ALWAYS_ALLOWED(TupleElement)
+    ALWAYS_ALLOWED(CaptureList)
+    ALWAYS_ALLOWED(Closure)
+    ALWAYS_ALLOWED(AutoClosure)
+    ALWAYS_ALLOWED(InOut)
+    ALWAYS_ALLOWED(DynamicType)
+    ALWAYS_ALLOWED(RebindSelfInConstructor)
+    ALWAYS_ALLOWED(BindOptional)
+    ALWAYS_ALLOWED(OptionalEvaluation)
+    ALWAYS_ALLOWED(ForceValue)
+    SOMETIMES_ALLOWED(Call)
+    ALWAYS_ALLOWED(PrefixUnary)
+    ALWAYS_ALLOWED(PostfixUnary)
+    ALWAYS_ALLOWED(Binary)
+    ALWAYS_ALLOWED(DotSyntaxCall)
+    ALWAYS_ALLOWED(ConstructorRefCall)
+    ALWAYS_ALLOWED(Load)
+    ALWAYS_ALLOWED(TupleShuffle)
+    ALWAYS_ALLOWED(InjectIntoOptional)
+    ALWAYS_ALLOWED(Coerce)
+    ALWAYS_ALLOWED(If)
+    ALWAYS_ALLOWED(Assign)
+    ALWAYS_ALLOWED(CodeCompletion)
+    ALWAYS_ALLOWED(EditorPlaceholder)
+
+    // Allow all errors and unchecked expressions so that we don't put errors
+    // on top of expressions that alrady have errors.
+    ALWAYS_ALLOWED(Error)
+    ALWAYS_ALLOWED(UnresolvedTypeConversion)
+    #define UNCHECKED_EXPR(ID, PARENT) ALWAYS_ALLOWED(ID)
+    #include "swift/AST/ExprNodes.def"
+
+    default:
```

Comment:
the `#include "swift/AST/ExprNodes.def"` above only defines cases for `UNCHECKED_EXPR` expressions

Reply:
Yeah, `ExprNodes.def` contains a bunch of types of nodes. `ExprNodes.def` defines macros that expand them all to nothing. I define the `UNCHECKED_EXPR` macro so that it expands to a `case`, and then include `ExprNodes.def`. So only the `UNCHECKED_EXPR`s end up contributing `case`s to the `switch`.

Reply:
Got it. Thanks for clarifying.